src/dagmc.cpp

```diff
@@ -37,189 +36,137 @@ const bool DAGMC_ENABLED = false;
 
 namespace openmc {
 
-const std::string DAGMC_FILENAME = "dagmc.h5m";
-
-namespace model {
-
-moab::DagMC* DAG;
-
-} // namespace model
+//==============================================================================
+// DAGMC Universe implementation
+//==============================================================================
 
+DAGUniverse::DAGUniverse(pugi::xml_node node) {
+  if (check_for_node(node, "id")) {
+    id_ = std::stoi(get_node_value(node, "id"));
+  } else {
+    fatal_error("Must specify the id of the DAGMC universe");
+  }
 
-std::string dagmc_file() {
-  std::string filename = settings::path_input + DAGMC_FILENAME;
-  if (!file_exists(filename)) {
-    fatal_error("Geometry DAGMC file '" + filename + "' does not exist!");
+  if (check_for_node(node, "filename")) {
+    filename_ = get_node_value(node, "filename");
+  } else {
+    fatal_error("Must specify a file for the DAGMC universe");
   }
-  return filename;
-}
 
-bool get_uwuw_materials_xml(std::string& s) {
-  std::string filename = dagmc_file();
-  UWUW uwuw(filename.c_str());
+  adjust_geometry_ids_ = false;
+  if (check_for_node(node, "auto_geom_ids")) {
+    adjust_geometry_ids_ = get_node_value_bool(node, "auto_geom_ids");
+  }
 
-  std::stringstream ss;
-  bool uwuw_mats_present = false;
-  if (uwuw.material_library.size() != 0) {
-    uwuw_mats_present = true;
-    // write header
-    ss << "<?xml version=\"1.0\"?>\n";
-    ss << "<materials>\n";
-    const auto& mat_lib = uwuw.material_library;
-    // write materials
-    for (auto mat : mat_lib) { ss << mat.second->openmc("atom"); }
-    // write footer
-    ss << "</materials>";
-    s = ss.str();
+  adjust_material_ids_ = false;
+  if (check_for_node(node, "auto_mat_ids")) {
+    adjust_material_ids_ = get_node_value_bool(node, "auto_mat_ids");
   }
 
-  return uwuw_mats_present;
+  initialize();
 }
 
-bool read_uwuw_materials(pugi::xml_document& doc) {
-  std::string s;
-  bool found_uwuw_mats = get_uwuw_materials_xml(s);
-  if (found_uwuw_mats) {
-    pugi::xml_parse_result result = doc.load_string(s.c_str());
-    if (!result) {
-      throw std::runtime_error{"Error reading UWUW materials"};
-    }
+DAGUniverse::DAGUniverse(const std::string& filename,
+                         bool auto_geom_ids,
+                         bool auto_mat_ids)
+: filename_(filename), adjust_geometry_ids_(auto_geom_ids), adjust_material_ids_(auto_mat_ids) {
+  // determine the next universe id
+  int32_t next_univ_id = 0;
+  for (const auto& u : model::universes) {
+    if (u->id_ > next_univ_id) next_univ_id = u->id_;
```

Comment:
will `u->id_` always be sorted smallest to largest?

Reply:
It isn't a requirement for `model::universes`, no. Are you concerned the next ID isn't correct?

Reply:
just wondering if it leads to any issues, like does it matter if the order isnt monotonically increasing?

Reply:
The IDs can be in any order here. Whenever we find one that's bigger, we update `next_univ_id`. If we happened to find the largest ID first, we'd only update that variable once.

Reply:
Are you happy with this implementation @makeclean?